File: 2013 ACM-ICPC North American Qualifier/erraticants.cs
using System;
using System.Collections.Generic;

public class Point : IEquatable<Point>
{
    public int X { get; set; }
    public int Y { get; set; }
    
    public Point(int x, int y)
    {
        X = x;
        Y = y;
    }
    
    public override bool Equals(object obj)
    {
        return Equals(obj as Point);
    }
    
    public bool Equals(Point other)
    {
        return other != null && X == other.X && Y == other.Y;
    }
    
    public override int GetHashCode()
    {
        return (X * 397) ^ Y;
    }
}

public class Program
{
    public static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        Console.ReadLine(); // Consume blank line
        
        for (int pathNum = 0; pathNum < n; pathNum++)
        {
            int s = int.Parse(Console.ReadLine());
            
            Dictionary<Point, HashSet<Point>> adj = new Dictionary<Point, HashSet<Point>>();
            Point currentPos = new Point(0, 0);
            Point startNode = new Point(0, 0);
            
            for (int i = 0; i < s; i++)
            {
                char dirChar = Console.ReadLine()[0];
                
                Point nextPos = new Point(currentPos.X, currentPos.Y);
                switch (dirChar)
                {
                    case 'N': nextPos.Y++; break;
                    case 'E': nextPos.X++; break;
                    case 'S': nextPos.Y--; break;
                    case 'W': nextPos.X--; break;
                }
                
                // Add bidirectional edges
                if (!adj.ContainsKey(currentPos))
                    adj[currentPos] = new HashSet<Point>();
                if (!adj.ContainsKey(nextPos))
                    adj[nextPos] = new HashSet<Point>();
                    
                adj[currentPos].Add(nextPos);
                adj[nextPos].Add(currentPos);
                
                currentPos = nextPos;
            }
            Point foodLocation = currentPos;
            
            // BFS
            Dictionary<Point, int> dist = new Dictionary<Point, int>();
            Queue<Point> queue = new Queue<Point>();
            int shortestPath = -1;
            
            dist[startNode] = 0;
            queue.Enqueue(startNode);
            
            while (queue.Count > 0)
            {
                Point u = queue.Dequeue();
                int d = dist[u];
                
                if (u.Equals(foodLocation))
                {
                    shortestPath = d;
                    break;
                }
                
                if (adj.ContainsKey(u))
                {
                    foreach (Point v in adj[u])
                    {
                        if (!dist.ContainsKey(v))
                        {
                            dist[v] = d + 1;
                            queue.Enqueue(v);
                        }
                    }
                }
            }
            
            Console.WriteLine(shortestPath);
            
            // Consume blank line if not last path
            if (pathNum < n - 1)
                Console.ReadLine();
        }
    }
}

File: 2013 ACM-ICPC North American Qualifier/erraticants.py
from collections import deque, defaultdict

def main():
    n = int(input().strip())
    input()  # Consume blank line
    
    for path_num in range(n):
        s = int(input().strip())
        
        adj = defaultdict(set)
        current_pos = (0, 0)
        start_node = (0, 0)
        
        for _ in range(s):
            dir_char = input().strip()
            
            x, y = current_pos
            if dir_char == 'N':
                next_pos = (x, y + 1)
            elif dir_char == 'E':
                next_pos = (x + 1, y)
            elif dir_char == 'S':
                next_pos = (x, y - 1)
            elif dir_char == 'W':
                next_pos = (x - 1, y)
            
            # Add bidirectional edges
            adj[current_pos].add(next_pos)
            adj[next_pos].add(current_pos)
            
            current_pos = next_pos
        
        food_location = current_pos
        
        # BFS
        dist = {}
        queue = deque()
        shortest_path = -1
        
        dist[start_node] = 0
        queue.append(start_node)
        
        while queue:
            u = queue.popleft()
            d = dist[u]
            
            if u == food_location:
                shortest_path = d
                break
            
            if u in adj:
                for v in adj[u]:
                    if v not in dist:
                        dist[v] = d + 1
                        queue.append(v)
        
        print(shortest_path)
        
        # Consume blank line if not last path
        if path_num < n - 1:
            input()

if __name__ == "__main__":
    main()